File: javasample/src/main/java/com/github/stephenvinouze/advancedrecyclerview/javasample/fragments/PaginationRecyclerFragment.java
package com.github.stephenvinouze.advancedrecyclerview.javasample.fragments;

import android.os.Bundle;
import android.os.Handler;
import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.swiperefreshlayout.widget.SwipeRefreshLayout;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;

import com.github.stephenvinouze.advancedrecyclerview.javasample.R;
import com.github.stephenvinouze.advancedrecyclerview.javasample.adapters.SamplePaginationAdapter;
import com.github.stephenvinouze.advancedrecyclerview.javasample.models.Sample;
import com.github.stephenvinouze.advancedrecyclerview.pagination.extensions.PaginationCallback;
import com.github.stephenvinouze.advancedrecyclerview.pagination.extensions.PaginationKt;

import java.util.List;

/**
 * Created by Stephen Vinouze on 06/11/2015.
 */
public class PaginationRecyclerFragment extends AbstractRecyclerFragment {

    private static final int PAGINATION_THRESHOLD = 5;

    private SwipeRefreshLayout refreshLayout;
    private SamplePaginationAdapter paginationAdapter;
    private final Handler handler = new Handler();

    @Nullable
    @Override
    public View onCreateView(@NonNull LayoutInflater inflater, ViewGroup container, Bundle savedInstanceState) {
        return inflater.inflate(R.layout.pagination_recycler_layout, container, false);
    }

    @Override
    public void onViewCreated(@NonNull View view, @Nullable Bundle savedInstanceState) {
        super.onViewCreated(view, savedInstanceState);

        paginationAdapter = new SamplePaginationAdapter();

        configureRecyclerView(view.findViewById(R.id.recycler_view));
        recyclerView.setAdapter(paginationAdapter);

        PaginationKt.enablePagination(
                recyclerView,
                PAGINATION_THRESHOLD,
                new PaginationCallback() {
                    @Override
                    public boolean isLoading() {
                        return paginationAdapter.isLoading();
                    }

                    @Override
                    public boolean hasAllItems() {
                        return false;
                    }

                    @Override
                    public void onLoad() {
                        populatePage(false, true);
                    }
                }
        );

        refreshLayout = view.findViewById(R.id.refresh_layout);
        refreshLayout.setOnRefreshListener(() -> populatePage(true, true));

        populatePage(true, false);
    }

    private void populatePage(boolean reload, boolean delayed) {
        paginationAdapter.setLoading(true);

        if (delayed) {
            handler.postDelayed(() -> loadPage(reload), 2000);
        } else {
            loadPage(reload);
        }
    }

    private void loadPage(boolean reload) {
        List<Sample> items = Sample.mockItems();
        if (reload) {
            paginationAdapter.setItems(items);
        } else {
            PaginationKt.appendItems(paginationAdapter, items);
        }

        paginationAdapter.setLoading(false);
        refreshLayout.setRefreshing(false);
    }
}

File: sample/src/main/java/com/github/stephenvinouze/advancedrecyclerview/sample/fragments/PaginationRecyclerFragment.kt
package com.github.stephenvinouze.advancedrecyclerview.sample.fragments

import android.os.Bundle
import android.os.Handler
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import androidx.fragment.app.Fragment
import androidx.recyclerview.widget.DefaultItemAnimator
import androidx.recyclerview.widget.LinearLayoutManager
import com.github.stephenvinouze.advancedrecyclerview.pagination.extensions.appendItems
import com.github.stephenvinouze.advancedrecyclerview.pagination.extensions.enablePagination
import com.github.stephenvinouze.advancedrecyclerview.sample.R
import com.github.stephenvinouze.advancedrecyclerview.sample.adapters.SamplePaginationAdapter
import com.github.stephenvinouze.advancedrecyclerview.sample.models.Sample
import kotlinx.android.synthetic.main.pagination_recycler_layout.*

/**
 * Created by Stephen Vinouze on 06/11/2015.
 */
class PaginationRecyclerFragment : Fragment() {

    private val handler = Handler()

    private val paginationAdapter: SamplePaginationAdapter by lazy {
        SamplePaginationAdapter()
    }

    override fun onCreateView(inflater: LayoutInflater, container: ViewGroup?, savedInstanceState: Bundle?): View? =
            inflater.inflate(R.layout.pagination_recycler_layout, container, false)

    override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
        super.onViewCreated(view, savedInstanceState)

        paginationRecyclerView.run {
            setHasFixedSize(true)
            layoutManager = LinearLayoutManager(context)
            itemAnimator = DefaultItemAnimator()
            adapter = paginationAdapter
            enablePagination(
                    isLoading = {
                        paginationAdapter.isLoading
                    },
                    hasAllItems = {
                        false
                    },
                    onLoad = {
                        populatePage(false, delayed = true)
                    })
        }

        refreshLayout.setOnRefreshListener { populatePage(true, delayed = true) }

        populatePage(true)
    }

    private fun populatePage(reload: Boolean, delayed: Boolean = false) {
        paginationAdapter.isLoading = true

        if (delayed) {
            handler.postDelayed({
                loadPage(reload)
            }, 2000)
        } else {
            loadPage(reload)
        }
    }

    private fun loadPage(reload: Boolean) {
        val items = Sample.mockItems()
        if (reload) {
            paginationAdapter.items = items.toMutableList()
        } else {
            paginationAdapter.appendItems(items)
        }

        paginationAdapter.isLoading = false
        refreshLayout.isRefreshing = false
    }
}
